### PR TITLE
Add option to print size on the left

### PR DESCRIPTION
**Pull request by bryceberger**

### CODE REVIEW

**Comment by bryceberger** [comment]
Ooh, just realized I forgot to add a test for combining left align and non-default precision. Should be implemented, just don't have a test for it. Will add tonight

File: src/render/tree/mod.rs [hunk 1]
```diff
@@ -199,74 +198,74 @@ impl Display for Tree {
...
         let theme = ui::get_theme();
-        let prune = self.context().prune;
-        let mut output = String::from("");
-
-        #[inline]
-        fn extend_output(output: &mut String, node: &Node, prefix: &str) {
-            output.push_str(&format!("{prefix}{node}\n"));
+        let size_left = self.context().size_left;
+
+        fn extend_output(
```
Comment (solidiquis) [inline src/render/tree/mod.rs:203]: Would these not benefit from being `#[inline]`?

File: src/render/tree/mod.rs [hunk 2]
```diff
@@ -199,74 +198,74 @@ impl Display for Tree {
...
+                writeln!(f, "")
+            } else {
+                // write!(f, "{prefix}")?;
+                node.display_size_right(f, prefix)?;
+                writeln!(f, "")
+            }
         }
 
-        #[inline]
         fn traverse(
```
Comment (solidiquis) [inline src/render/tree/mod.rs:220]: Same comment as before about inlining

**Review by solidiquis — COMMENTED** [review]
Thanks for whipping this up. I won't be able to take a deeper look until later in the week as it's a work week, but I left some questions.

It also bothers me a bit that the root directory's disk usage isn't aligned with everyone else's. Would it be too difficult to offset that as well? Or perhaps there are other considerations that I'm missing.

File: src/render/tree/mod.rs (hunk 1 above)
Comment (bryceberger) [reply src/render/tree/mod.rs:203]: ...maybe. `#[inline]` is, at best, a [suggestion](https://nnethercote.github.io/perf-book/inlining.html) -- this function is so tiny it's almost guaranteed to be inclined (in release builds, at least).

Even if not inlined without the attribute (but inlined with it), I don't think you'd be able to see a measurable performance change -- the variance of the io-bound traversal would probably dwarf the change.

File: src/render/tree/mod.rs (hunk 2 above)
Comment (bryceberger) [reply src/render/tree/mod.rs:220]: This function is directly recursive -- it cannot be inlined without some transformation that removes the recursion.

Assuming that such a transformation exists (and the compiler finds it), it's only called in one place (aside from the recursive case). The choice of whether or not to inline would not affect performance -- a single function call is not measureable.

**Comment by bryceberger** [comment]
> It also bothers me a bit that the root directory's disk usage isn't aligned with everyone else's.

Not sure what you mean by this, everything is aligned in the output for me.

Maybe you were talking about the string literals in the `size_left.rs` test file? `indoc` didn't work because I wanted some indentation, and there isn't leading indentation on the first line because of the `.trim()` on line 25 of `tests/utils/mod.rs`

File: src/render/tree/mod.rs (hunk 1 above)
Comment (solidiquis) [reply src/render/tree/mod.rs:203]: that is fair

**Comment by solidiquis** [comment]
Ahh yes that was what I was referring to. In that case fantastic. I'll do a deeper review tonight! Also I did not mean to edit your comment above. I did not know that was even possible.

File: src/render/disk_usage.rs [hunk 3]
```diff
@@ -94,108 +94,111 @@ impl AddAssign<&Self> for FileSize {
     }
 }
 
-impl Display for FileSize {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+impl FileSize {
+    pub fn format(&self, align: bool) -> String {
```
Comment (solidiquis) [inline src/render/disk_usage.rs:98]: Doc comment here pls

File: src/render/tree/node.rs [hunk 4]
```diff
@@ -314,45 +312,68 @@ impl From<(&DirEntry, &Context)> for Node {
     }
 }
 
+#[derive(Copy, Clone, Default)]
+enum SizeLocation {
```
Comment (solidiquis) [inline src/render/tree/node.rs:316]: Doc comment here pls

File: src/render/tree/node.rs [hunk 5]
```diff
@@ -314,45 +312,68 @@ impl From<(&DirEntry, &Context)> for Node {
...
+    #[default]
+    Right,
+    Left,
+}
+
+impl SizeLocation {
```
Comment (solidiquis) [inline src/render/tree/node.rs:322]: Doc comments for these methods please

File: src/render/tree/node.rs [hunk 6]
```diff
@@ -314,45 +312,68 @@ impl From<(&DirEntry, &Context)> for Node {
...
+}
+
 impl Display for Node {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        self.display(f, SizeLocation::default(), "")
+    }
+}
+
+impl Node {
+    fn display(&self, f: &mut Formatter, size_loc: SizeLocation, prefix: &str) -> fmt::Result {
```
Comment (solidiquis) [inline src/render/tree/node.rs:346]: Doc comment here pls

File: src/render/tree/node.rs [hunk 7]
```diff
@@ -314,45 +312,68 @@ impl From<(&DirEntry, &Context)> for Node {
...
+        match size_loc {
+            SizeLocation::Right => {
+                write!(f, "{prefix}{icon:<icon_padding$}{styled_name} {size}",)
+            }
+            SizeLocation::Left => {
+                write!(f, "{size} {prefix}{icon:<icon_padding$}{styled_name}",)
+            }
+        }
+    }
+    pub fn display_size_right(&self, f: &mut Formatter, prefix: &str) -> fmt::Result {
```
Comment (solidiquis) [inline src/render/tree/node.rs:373]: Doc comment here pls

File: src/render/tree/node.rs [hunk 8]
```diff
@@ -314,45 +312,68 @@ impl From<(&DirEntry, &Context)> for Node {
...
+        self.display(f, SizeLocation::Right, prefix)
+    }
+    pub fn display_size_left(&self, f: &mut Formatter, prefix: &str) -> fmt::Result {
```
Comment (solidiquis) [inline src/render/tree/node.rs:376]: Doc comment here pls

**Review by solidiquis — CHANGES_REQUESTED** [review]
Everything looks great! Could you just go through and add doc comments to all the places I noted? I know that it's pretty obvious what a lot of these things are doing but it's still something I wish to enforce to be thorough.

**Review by solidiquis — APPROVED** [review]

**Comment by solidiquis** [comment]
Looks good to me! Thanks again for the contribution @bryceberger. For future reference if you wouldn't mind an update to the usage section of the README whenever new cli argument/options are introduced that'd be awesome. But I got it this time around :]

**Comment by solidiquis** [comment]
And wow this looks so nice :D

**Comment by solidiquis** [comment]
I included a screenshot in the README and decided to add `--size-left` as a default in my erdtree config. Thanks again!

**Pull request merged**